### PR TITLE
esp8266: Factory function for SPI

### PR DESCRIPTION
**Pull request by deshipu**
Rename machine.SPI to machine.SoftSPI and create a factory function machine.SPI that will call either machine.SoftSPI or machine.HSPI depending on the parameters.

### CODE REVIEW

File: esp8266/modmachine.c [hunk 1]
```diff
@@ -225,6 +225,46 @@ STATIC mp_obj_t machine_enable_irq(mp_obj_t state_in) {
...
+        { MP_QSTR_sck, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_mosi, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_miso, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args),
+                     allowed_args, args);
+    switch (args[ARG_index].u_int) {
+        case -1: {
+            // create and return SoftSPI
```
Comment (deshipu) [inline esp8266/modmachine.c]: I need a little help here. I looked around, but couldn't find a way to create a python object from within C and to pass all the positional and keyword arguments, except index, to it.

**Comment by pfalcon** [comment]
Well, so the idea should be that this factory function is really a standalone demultiplexer - it should do minimum work it needs to be done to get its job, and that's looking at the first positional argument, or "id" keyword argument, to decide which constructor to call with the rest of params. Thanks for splitting changes in 2 commits, but they should be in opposite order - real job first, cosmetic stuff later, and there's always if the latter is really needed, so I'd wait for first patch to be merged before going for 2nd (and would ask if it's needed).

**Comment by deshipu** [comment]
I don't understand. How do you expect me to do the commits in the opposite order, when there is a name conflict? I had to move the conflicting name out of the way first.

As for real job first -- I did what I could deduce from the existing code. There is no documentation on this stuff, so I figured I will do as much as I can, and then ask for help about the parts I'm missing. Now, the question is, can you (or anybody else) help with this or not? If not, I will just close both pull requests and forget about the whole thing.

**Comment by pfalcon** [comment]
And yes, signature difference of make_new() "special" method and normal method is "unfortunate" (originally based on "be more efficient" idea, though I wonder if may need to revisit that). So, if you'd go for making this new "SPI" multiplexer a type, just for the sake of .make_new method, I'd find it acceptable (can be refactored later). Otherwise, there's definitely a function to convert between argument passing conventions.

**Comment by pfalcon** [comment]
Well, you can see argcheck.c, mp_arg_parse_all_kw_array(). That converts make_new() style args to normal args, and as can be seen, pretty trivial. Opposite conversion wouldn't be hard either.

@dpgeorge : Maybe we need a way to create a normal method with "raw" (make_new()-style) argument passing? ;-).

**Comment by pfalcon** [comment]
> I don't understand. How do you expect me to do the commits in the opposite order, when there is a name conflict? I had to move the conflicting name out of the way first.

C and Python names live in different worlds, ones aren't tied to other. What we're trying to do is to give way for *Python* name, so changing C names aren't really *required* (it may be nice though - or not, to be decided further).

> Now, the question is, can you (or anybody else) help with this or not?

Helping is what I'm trying to do, and I assume if you went this far, you want to learn more things (which I explain in as little time as I have, so they're cursory, but the other alternative is it hanging a long in the queue). But that's indeed "hard" side of things for, so feel free to say any time "I've done what could, can you take it from there?"

**Comment by deshipu** [comment]
I updated the original pull request for HSPI2 with this.

**Pull request closed without merging**